partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelJoinOuterRight.kt

```diff
@@ -0,0 +1,82 @@
+package org.partiql.eval.internal.operator.rel
+
+import org.partiql.eval.internal.Environment
+import org.partiql.eval.internal.Record
+import org.partiql.eval.internal.helpers.TypesUtility.toRuntimeType
+import org.partiql.eval.internal.helpers.ValueUtility.isTrue
+import org.partiql.eval.internal.helpers.toNull
+import org.partiql.eval.internal.operator.Operator
+import org.partiql.plan.Rel
+import org.partiql.value.PartiQLValueExperimental
+
+internal class RelJoinOuterRight(
+    private val lhs: Operator.Relation,
+    private val rhs: Operator.Relation,
+    private val condition: Operator.Expr,
+    lhsType: Rel.Type
+) : RelPeeking() {
+
+    @OptIn(PartiQLValueExperimental::class)
+    private val lhsPadded = Record(
+        Array(lhsType.schema.size) { lhsType.schema[it].type.toRuntimeType().toNull().invoke() }
+    )
+
+    private lateinit var env: Environment
+    private lateinit var iterator: Iterator<Record>
+
+    override fun openPeeking(env: Environment) {
+        this.env = env
+        rhs.open(env)
+        iterator = implementation()
+    }
+
+    override fun peek(): Record? {
+        return when (iterator.hasNext()) {
+            true -> iterator.next()
+            false -> null
+        }
+    }
+
+    override fun closePeeking() {
+        lhs.close()
+        rhs.close()
+        iterator = emptyList<Record>().iterator()
+    }
+
+    /**
+     * RIGHT OUTER JOIN (LATERAL)
+     *
+     * Algorithm:
+     * ```
+     * for rhsRecord in rhs:
+     *   for lhsRecord in lhs(rhsRecord):
+     *     if (condition matches):
+     *       conditionMatched = true
+     *       yield(lhsRecord + rhsRecord)
+     *   if (!conditionMatched):
+     *     yield(NULL_RECORD + rhsRecord)
+     * ```
+     *
+     * Another Note: For some databases, RIGHT OUTER JOIN cannot be used with LATERAL. This should be taken care of
+     * at planning time. TODO: Should we support LATERAL here? Can we deal with this at planning time?
+     */
+    @OptIn(PartiQLValueExperimental::class)
+    private fun implementation() = iterator {
+        for (rhsRecord in rhs) {
+            var rhsMatched = false
+            lhs.open(env.push(rhsRecord))
```

Comment:
I found this feature kind of counter-intuitive (as in SQL most stuff reads from left to right), and arguably incompatible with the PartiQL Semantics (as in PartiQL the "inner loop" usually can access binding environment in the "outer loop")

I believe that in set theory:`A LEFT OUTER JOIN B` is the same as `B RIGHT OUTER JOIN A`. 

In SQL implementation, the difference might be the order of attribute in the produced tuple....

I think this implementation is trying to achieve the same result.... But I am NOT convinced that we should allow the left hand side relation to have access to the binding env of rhs relation....